vslib/src/SwitchStateBaseMACsec.cpp

```diff
@@ -0,0 +1,831 @@
+{
+    SWSS_LOG_ENTER();
+
+    std::vector<sai_attribute_t> attrs;
+    sai_attribute_t attr;
+
+    // MACsec flow => ACL entry => ACL table
+    // MACsec flow => MACsec SC
+    // (ACL table & MACsec SC) => port
+
+    // Find ACL Entry
+    attr.id = SAI_ACL_ENTRY_ATTR_ACTION_MACSEC_FLOW;
+    attr.value.aclaction.parameter.oid = macsecFlowId;
+    attr.value.aclaction.enable = true;
+    std::vector<sai_object_id_t> aclEntryIds;
+    findObjects(SAI_OBJECT_TYPE_ACL_ENTRY, attr, aclEntryIds);
+
+    if (aclEntryIds.empty())
+    {
+        // No ACL Entry is bound to the flow
+        SWSS_LOG_DEBUG(
+            "Cannot find corresponding ACL entry for the flow %s",
+            sai_serialize_object_id(macsecFlowId).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    // Find ACL Table
+    sai_object_id_t aclTableId;
+
+    // All ACL entries with same MACsec flow correspond one ACL Table
+    // Because an ACL Table corresponds one MACsec Port
+    // And a flow never belongs to two MACsec Port
+    if (getACLTable(aclEntryIds.front(), aclTableId) != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR(
+            "Cannot find corresponding ACL table for the entry %s",
+            sai_serialize_object_id(aclEntryIds.front()).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    // Find MACsec SC
+    attr.id = SAI_MACSEC_SC_ATTR_FLOW_ID;
+    attr.value.oid = macsecFlowId;
+    std::vector<sai_object_id_t> macsecScIds;
+    findObjects(SAI_OBJECT_TYPE_MACSEC_SC, attr, macsecScIds);
+
+    // One MACsec SC will only belong to one MACsec flow
+    // Meanwhile one MACsec flow will just belong to one port
+    if (macsecScIds.empty())
+    {
+        SWSS_LOG_ERROR(
+            "Cannot find corresponding MACsec SC for the flow %s",
+            sai_serialize_object_id(macsecFlowId).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    attrs.resize(1);
+    attrs[0].id = SAI_MACSEC_SC_ATTR_MACSEC_DIRECTION;
+
+    CHECK_STATUS(get(SAI_OBJECT_TYPE_MACSEC_SC, macsecScIds.front(), static_cast<uint32_t>(attrs.size()), attrs.data()));
+
+    auto direction = attrs[0].value.s32;
+
+    // Find port
+    attrs.resize(1);
+    attrs[0].id = (direction == SAI_MACSEC_DIRECTION_EGRESS) ? SAI_PORT_ATTR_EGRESS_MACSEC_ACL : SAI_PORT_ATTR_INGRESS_MACSEC_ACL;
+    attrs[0].value.oid = aclTableId;
+
+    std::vector<sai_object_id_t> port_ids;
+
+    findObjects(SAI_OBJECT_TYPE_PORT, attrs.back(), port_ids);
+
+    if (port_ids.empty())
+    {
+        SWSS_LOG_ERROR(
+            "Cannot find corresponding port for the ACL table %s",
+            sai_serialize_object_id(aclTableId).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+    else if (port_ids.size() > 1)
+    {
+        SWSS_LOG_ERROR(
+            "Duplicated ports for the ACL table %s",
+            sai_serialize_object_id(aclTableId).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    portId = port_ids.front();
+
+    return SAI_STATUS_SUCCESS;
```

Comment:
better and smaller:
if (port.size() == 1)
{
portId = port_ids.front();
return SUCCESS
}

SWSS_LOG_ERROR("expected 1 port, but got %zu", )
return FAILURE;